### PR TITLE
ENH: add DataFrame.is_unique method

### PR DESCRIPTION
**Pull request by topper-123**
Currently, to see columns with unique values, we have to do `df.apply(lambda x: x.is_unique)`, which is a bit awkward.

IMO checking for unique columns in a DataFrame is a common enough need to justify a method for that specific purpose.

Xref: #11948

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -1038,16 +1038,19 @@ def nunique(self, dropna: bool = True) -> int:
             n -= 1
         return n
 
+    def _is_unique(self) -> bool:
```
Comment (jreback) [inline pandas/core/base.py:1041]: i don't think this necessary

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5199,6 +5199,34 @@ def drop_duplicates(
...
+
+        Returns
+        -------
+        Series
+
+        See Also
+        --------
+        DataFrame.duplicated : Indicate duplicate rows.
+        """
+        if subset is not None:
```
Comment (jreback) [inline pandas/core/frame.py:5221]: don't separate the paths, just
```self = self[subset]```

File: pandas/core/frame.py [hunk 3]
```diff
@@ -5199,6 +5199,34 @@ def drop_duplicates(
...
+            subset = subset if is_list_like(subset) else [subset]
+            return self.loc[:, subset].is_unique()
+
+        if len(self.columns):
+            return self.apply(Series._is_unique)
```
Comment (jreback) [inline pandas/core/frame.py:5226]: this should *just* work with
```return self.apply(lambda x: x.is_unique())```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/frame.py:5221]: Yeah, changed.

File: pandas/core/frame.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/frame.py]: Seemed to be the the same, but I guess someone could have subclassed `_constructor_sliced`, which could make a difference.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -5199,6 +5199,34 @@ def drop_duplicates(
...
+        Returns
+        -------
+        Series
+
+        See Also
+        --------
+        DataFrame.duplicated : Indicate duplicate rows.
+        """
+        if subset is not None:
+            subset = subset if is_list_like(subset) else [subset]
```
Comment (jreback) [inline pandas/core/frame.py:5222]: can you maybe_make_list

File: pandas/core/frame.py [hunk 5]
```diff
@@ -5199,6 +5199,34 @@ def drop_duplicates(
...
+            self = self[subset]
+
+        if len(self.columns):
+            return self.apply(lambda x: x.is_unique)
```
Comment (jreback) [inline pandas/core/frame.py:5226]: why does this not work on empties?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
For a DataFrame, `is_unique` could also be interpreted as whether the DataFrame has all unique rows?

File: pandas/core/frame.py (hunk 4 above)
Comment (topper-123) [reply pandas/core/frame.py:5222]: Ok

File: pandas/core/frame.py (hunk 5 above)
Comment (topper-123) [reply pandas/core/frame.py:5226]: A dataframe with no columns doesn't call the inner func in apply (here `lambda x: x.is_unique`), so the output (empty) series has to guess its own dtype. It guesses `float64`, which is wrong, so we have to special-case it.

**Comment by topper-123** [comment]
@jorisvandenbossche that's true, and could be fixes by having an `axis` parameter. I didn't ínclude it, because `duplicated` doesn't have one and worried that doing `self.T.is_unique()` would change the types for the values and could give wrong results. Not sure about that one, actually.

**Comment by jreback** [comment]
@topper-123 this was close, can you merge master and move the note to 1.3

**Comment by topper-123** [comment]
Ok, done.

**Comment by mroeschke** [comment]
Looks pretty good overall. Should there be a corresponding `Series.is_unique` method as well? There is a `Series.duplicated` method (as your referenced `duplicated` earlier)

**Comment by mroeschke** [comment]
Thanks for the PR, but it appears that it's gone stale. Please let us know if you'd like to continue by merging master and targeting for 1.4.

**Pull request closed without merging**